logic/transactions/pool.js

```diff
@@ -0,0 +1,922 @@
+ */
+TransactionPool.prototype.addFromPeer = function (transactions, broadcast, cb) {
+	if (!Array.isArray(transactions)) {
+		transactions = [transactions];
+	}
+	async.eachSeries(transactions, function (transaction, cb) {
+		__private.addToUnverified(transaction, broadcast, cb);
+	}, function (err) {
+		return setImmediate(cb, err);
+	});
+};
+
+/**
+ * Adds transactions to verified.ready pool list.
+ * @implements {__private.addToPoolList}
+ * @implements {delete}
+ * @param {transaction} transactions
+ * @param {function} cb - Callback function.
+ * @return {setImmediateCallback} error | cb
+ */
+TransactionPool.prototype.addReady = function (transactions, cb) {
+	if (!Array.isArray(transactions)) {
+		transactions = [transactions];
+	}
+	var resetReceivedAt = new Date();
+	async.eachSeries(transactions, function (transaction, eachSeriesCb) {
+		self.delete(transaction);
+		transaction.receivedAt = resetReceivedAt;
+		__private.addToPoolList(transaction, pool.verified.ready, eachSeriesCb);
+	}, function (err) {
+		return setImmediate(cb, err);
+	});
+};
+
+/**
+ * Creates and adds signature to multisignature transaction.
+ * @implements {__private.createSignature}
+ * @param {String} transactionId transaction id
+ * @param {String} secret secret
+ * @param {function} cb - Callback function.
+ * @return {setImmediateCallback} error | cb
+ */
+TransactionPool.prototype.addSignature = function (transactionId, secret, cb) {
+	var multisignatureTransaction = pool.verified.pending.transactions[transactionId];
+
+	if (multisignatureTransaction === undefined) {
+		library.logger.error(['Failed to add signature to multisignature. Transaction', transactionId, 'not in pool'].join(' '));
+		return setImmediate(cb, 'Transaction not in pool');
+	}
+
+	// TODO: replace with checkSignature to reflect API 1.0.18 functionality
+	__private.createSignature(multisignatureTransaction, secret, function (err, signature) {
+		if (err) {
+			return setImmediate(cb, err);
+		}
+		if (multisignatureTransaction.signatures.indexOf(signature) !== -1) {
+			library.logger.error(['Transaction already signed:', transactionId].join(' '));
+			return setImmediate(cb, 'Transaction already signed');
+		}
+	
+		multisignatureTransaction.signatures.push(signature);
+		return setImmediate(cb);
+	});
+};
+
+/**
+ * Deletes transaction from pool list.
+ * @implements {__private.delete}
+ * @param {string} id transaction id
+ * @return {Array} names of cleared lists
+ */
+TransactionPool.prototype.delete = function (id) {
+	var clearedList = [];
+	var poolList = ['unverified','pending','ready'];
+
+	[pool.unverified, pool.verified.pending, pool.verified.ready].forEach(function (list, index) {
+		if (__private.delete(id, list)) {
+			clearedList.push(poolList[index]);
+		}
+	});
+	if (clearedList.length > 0) {
+		if (clearedList.length > 1) {
+			library.logger.debug(['Cleared duplicated transaction in pool list:', clearedList, 'transaction id:', id].join(' '));
```

Comment:
Whjat duplicated transactions? It should never hit that, if it hits here - logic is wrong